core/src/wallet/bitcoin/synchronizers/BlockchainExplorerAccountSynchronizer.cpp

```diff
@@ -95,5 +135,191 @@ namespace ledger {
+                    // and go though all inputs to find foreign address
+                    replaceableTxs.emplace_back(txWithHash.second.first);
+                    for (const auto& input : txWithHash.second.first.inputs) {
+                        if (input.address.hasValue() &&
+                            !buddy->account->getKeychain()->contains(input.address.getValue())) {
+                            foreignAddresses.insert(input.address.getValue());
+                        }
+                    }
+                } else {
+                    notReplaceableTxs.emplace_back(txWithHash.second.first);
+                }
+            }
+            // Use the list of addresses to get all mempool transaction for those addresses
+            return getMempoolTransactions(buddy,
+                    std::vector<std::string>(foreignAddresses.begin(), foreignAddresses.end()))
+            .map<Unit>(buddy->account->getContext(), [=] (const auto& foreignTxs) {
+                // Do another round of filtering on replaceable transaction vs foreign transactions
+                TransactionMap filteredTxs;
+                std::list<Transaction> txs(replaceableTxs.begin(), replaceableTxs.end());
+                txs.insert(txs.end(), foreignTxs.begin(), foreignTxs.end());
+                filterReplacedTransaction(txs, filteredTxs);
+                // Insert everything in database
+                soci::session sql(buddy->account->getWallet()->getDatabase()->getPool());
+                soci::transaction tr(sql);
+                // Insert not replaceable transactions
+                for (const auto& tx : notReplaceableTxs) {
+                    buddy->account->putTransaction(sql, tx);
+                }
+                // Insert replaceable transaction
+                for (const auto& entry : filteredTxs) {
+                    buddy->account->putTransaction(sql, entry.second.first);
+                }
+                tr.commit();
+                return unit;
+            });
+        }
+
+        Future<Unit> BlockchainExplorerAccountSynchronizer::synchronizeMempool(
+                const std::shared_ptr<AbstractBlockchainExplorerAccountSynchronizer<BitcoinLikeAccount, BitcoinLikeAddress, BitcoinLikeKeychain, BitcoinLikeBlockchainExplorer>::SynchronizationBuddy> &buddy) {
+            auto bitcoinBuddy = std::static_pointer_cast<BitcoinSynchronizationBuddy>(buddy);
+            return resolveMempool(bitcoinBuddy);
+        }
+
+        Future<Unit> BlockchainExplorerAccountSynchronizer::recoverFromFailedSynchronization(
+                const std::shared_ptr<CommonBuddy> &commonBuddy) {
+            auto buddy = std::static_pointer_cast<BitcoinSynchronizationBuddy>(commonBuddy);
+            return Future<Unit>::async(buddy->account->getContext(), [buddy] () {
+                soci::session sql(buddy->account->getWallet()->getDatabase()->getPool());
+                for (const auto& tx : buddy->previousMempool) {
```

Comment:
Why do you want to insert transactions from mempool? We should drop them each time we want to do synchronizetion, no?

Reply:
For the specific case of recoverFromFailedSynchronization:
I don't want a fail transaction to change the database  (it can fail because internet is down for example). But since I'm deleting all mempool transaction at the beginning of the synchronization, I restore the database to its previous state (artificially restore the database to its previous state in fact).

recoverFromFailedSynchronization should be only call if the synchronization process failed.